yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+   */
+  private async _getIndexedLowWitness<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    value: Fr,
+    T: { fromBuffer: (buffer: Buffer) => T },
+  ): Promise<LowWitness<T>> {
+    // Get low public data witness
+    const { preimage: lowWitness, index } = await this.getLowInfo(treeId, value, T);
+    // We check if the value is already in the tree
+    const isUpdate: boolean = lowWitness.getNextKey() === value.toBigInt();
+    if (isUpdate) {
+      const lowSiblingPath =
+        this.treeMap.get(treeId)!.getSiblingPath(index) ?? (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+      const result = {
+        update: true,
+        lowSiblingPath: lowSiblingPath,
+        preimage: lowWitness,
+        index,
+      };
+      return result;
+    }
+    // This is an insertion of a new entry
+    const lowSiblingPath =
+      this.treeMap.get(treeId)!.getSiblingPath(index) ?? (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+    const oldLowWitness = T.fromBuffer(lowWitness.toBuffer());
+
+    // Build Result struct
+    const result = {
+      update: false,
+      lowSiblingPath: lowSiblingPath,
+      preimage: oldLowWitness,
+      index,
+    };
+
+    return result;
+  }
+}
+
+/****************************************************/
+/****** Some useful Structs and Enums **************/
+/****************************************************/
+enum TreeType {
+  LEAF,
+  BRANCH,
+  EMPTY,
+}
+
+type Leaf = {
+  tag: TreeType.LEAF;
+  value: Fr;
+};
+type Branch = {
+  tag: TreeType.BRANCH;
+  leftTree: Tree;
+  rightTree: Tree;
+};
+type Empty = {
+  tag: TreeType.EMPTY;
+};
+
+type Tree = Leaf | Branch | Empty;
+
+enum SiblingStatus {
+  MEMBER,
+  NONMEMBER,
+  ERROR,
+}
+
+type AccumulatedSiblingPath = {
+  path: Fr[];
+  status: SiblingStatus;
+};
+
+/****************************************************/
+/****** Some Helpful Constructors for Trees  ********/
+/****************************************************/
+const Branch = (left: Tree, right: Tree): Branch => ({
+  tag: TreeType.BRANCH,
+  leftTree: left,
+  rightTree: right,
+});
+
+const Leaf = (value: Fr): Leaf => ({
+  tag: TreeType.LEAF,
+  value,
+});
+
+const Empty = (): Empty => ({
+  tag: TreeType.EMPTY,
+});
+
+/****************************************************/
+/****** The EphemeralAvmTree Class *****************/
+/****************************************************/
+
+/**
+ * This is a recursively defined tree that has leaves at different heights
+ * It is seeded by an existing merkle treeDb for which it derives a frontier
+ * It is intended to be a lightweight tree that contains only the necessary information to suppport appends or updates
+ */
+export class EphemeralAvmTree {
+  public tree: Tree;
+  public zeroHashes: Fr[];
+  public leafCount;
+  public frontier: Fr[] = [];
```

Comment:
can some of the be private

Reply:
yep i had a bunch of these for debugging